torch/lib/c10d/ProcessGroupGloo.cpp

```diff
@@ -606,6 +859,74 @@ class AsyncAllreduceCUDAWork : public AsyncAllreduceWork {
   std::vector<at::cuda::CUDAEvent> events;
 };
 
+class AsyncSparseAllreduceCUDAWork : public AsyncSparseAllreduceWork {
+ public:
+  AsyncSparseAllreduceCUDAWork(
+      const std::shared_ptr<gloo::Context>& context,
+      std::vector<at::Tensor>& inputs,
+      uint32_t tag)
+      : AsyncSparseAllreduceWork(context, inputs, tag) {
+    initializeStreamsEvents(inputs, streams, events);
+
+    // Kick off copy from CUDA tensors to pinned CPU tensors.
+    tmp.reserve(inputs.size());
+    at::cuda::OptionalCUDAStreamGuard guard;
+    for (size_t i = 0; i < inputs.size(); i++) {
+      guard.reset_stream(streams[i]);
+      auto input_coalesced = inputs[i].coalesce();
+      auto input_indices = input_coalesced.indices();
+      auto input_values = input_coalesced.values();
+      auto indices = pinnedLike(input_indices).copy_(input_indices, true);
+      auto values = pinnedLike(input_values).copy_(input_values, true);
+      tmp.push_back(at::sparse_coo_tensor(indices, values, inputs[i].sizes()));
+    }
+  }
+
+  void run() override {
+    // Synchronize with copy operations.
+    at::cuda::OptionalCUDAGuard device_guard;
+    for (size_t i = 0; i < inputs.size(); i++) {
+      device_guard.set_index(inputs[i].device().index());
+      AT_CUDA_CHECK(cudaStreamSynchronize(streams[i]));
+    }
+
+    // Run allreduce on host side tensors.
+    auto output = allreduce(tmp);
+
+    // Kick off copy back to the CUDA tensors.
+    at::cuda::OptionalCUDAStreamGuard stream_guard;
+    for (size_t i = 0; i < inputs.size(); i++) {
+      stream_guard.reset_stream(streams[i]);
+      auto indices = output.indices().to(
+          inputs[i].device(),
+          /*non_blocking=*/true,
+          /*copy=*/true);
+      auto values = output.values().to(
+          inputs[i].device(),
+          /*non_blocking=*/true,
+          /*copy=*/true);
+
+      outputs.push_back(
```

Comment:
If we are not writing back to the input tensors for sparse use case, we should edit Python docs accordingly. BTW, we can still write to input tensors using `tensor.set_()`?